File: scripts/office-quality/docx-benchmark.mjs
import { createHash } from 'node:crypto';
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { resolve } from 'node:path';
import { pathToFileURL } from 'node:url';
import { fetchAsset } from './asset-cache.mjs';
import { download } from './download.mjs';
import { mapPool, validatePlan } from './plan.mjs';
import { validateComparison } from './results.mjs';
import { CORPUS_ORIGIN } from './samples.mjs';

export const CHANNELS = ['published', 'commit', 'libreoffice'];
export const TRIALS = 5;
export const METHOD = 'docx-cli-page1-v1';
export const digest = (bytes) => createHash('sha256').update(bytes).digest('hex');

export function docxShards(plan) {
  const samples = plan.samples.filter((sample) => sample.format === 'docx');
  const shards = Array.from({ length: Math.min(4, samples.length) }, () => []);
  const weights = shards.map(() => 0);
  for (const sample of [...samples].sort((a, b) =>
    b.metadata.reference.pages - a.metadata.reference.pages || a.id.localeCompare(b.id))) {
    const index = weights.indexOf(Math.min(...weights));
    shards[index].push(sample.id);
    weights[index] += sample.metadata.reference.pages;
  }
  return shards;
}

function fail(message) {
  throw new Error(`Invalid DOCX benchmark: ${message}`);
}

function validateTiming(timing) {
  if (timing?.status === 'failed') {
    if (typeof timing.error !== 'string' || !timing.error.trim() || 'elapsed_ms' in timing)
      fail('invalid timing failure');
    return;
  }
  if (timing?.status !== 'ok' || !Array.isArray(timing.elapsed_ms) ||
      timing.elapsed_ms.length !== TRIALS ||
      !timing.elapsed_ms.every((value) => Number.isFinite(value) && value > 0))
    fail('invalid timing trials');
}

export function timingSummary(samples, format = 'docx') {
  const rows = samples.filter((sample) => sample.format === format);
  for (const row of rows) {
    if (!row.native_timings || Object.keys(row.native_timings).length !== CHANNELS.length)
      fail('missing timing channels');
    for (const channel of CHANNELS) validateTiming(row.native_timings[channel]);
  }
  const common = rows.filter((row) => CHANNELS.every((channel) => row.native_timings[channel].status === 'ok'));
  return {
    total: rows.length,
    common: common.length,
    channels: Object.fromEntries(CHANNELS.map((channel) => [channel, {
      successful: rows.filter((row) => row.native_timings[channel].status === 'ok').length,
      mean_ms: common.length ? common.reduce((sum, row) =>
        sum + row.native_timings[channel].elapsed_ms.reduce((a, b) => a + b, 0) / TRIALS, 0) / common.length : null,
    }])),
  };
}

export function mergeDocxBenchmarks(plan, report, inputs, planHash) {
  const shards = docxShards(plan);
  if (!/^[a-f0-9]{40}$/.test(plan.docx_published_source_sha ?? '')) fail('missing published source');
  if (!Array.isArray(inputs) || inputs.length !== shards.length || !shards.length)
    fail('missing or unexpected shard reports');
  const seen = new Set();
  const results = new Map();
  let identity;
  for (const part of inputs) {
    if (part?.schema_version !== 1 || part.plan_sha256 !== planHash ||
        !Number.isInteger(part.shard) || !shards[part.shard] || seen.has(part.shard))
      fail('duplicate shard or mismatched plan');
    seen.add(part.shard);
    const config = part.benchmark;
    if (config?.method !== METHOD || config.trials !== TRIALS || config.dpi !== 96 ||
        !/^\d+(?:\.\d+){2,3}$/.test(config.libreoffice_version ?? '') ||
        !/^[a-f0-9]{64}$/.test(config.fonts_sha256 ?? '') ||
        config.published_version !== plan.versions.docx ||
        !config.builds || Object.keys(config.builds).length !== 2 ||
        config.builds?.published?.source_sha !== plan.docx_published_source_sha ||
        config.builds?.commit?.source_sha !== plan.source_sha)
      fail('invalid executable or measurement identity');
    for (const build of Object.values(config.builds)) {
      if (!/^[a-f0-9]{64}$/.test(build.binary_sha256 ?? '') ||
          !/^[a-f0-9]{64}$/.test(build.harness_sha256 ?? '')) fail('invalid build hashes');
    }
    if (config.builds.published.harness_sha256 !== config.builds.commit.harness_sha256 ||
        config.builds.published.rustc !== config.builds.commit.rustc ||
        config.builds.published.profile !== config.builds.commit.profile)
      fail('native builds use different harnesses or compilers');
    const serialized = JSON.stringify(config);
    if (identity && identity !== serialized) fail('shards use different benchmark environments');
    identity = serialized;
    if (!Array.isArray(part.samples) || part.samples.length !== shards[part.shard].length)
      fail('missing samples');
    for (const row of part.samples) {
      if (!shards[part.shard].includes(row?.id) || results.has(row.id))
        fail('duplicate, unexpected, or misplaced sample');
      const sample = plan.samples.find((sample) => sample.id === row.id);
      if (row.source_sha256 !== sample.metadata.source.sha256) fail('source hash mismatch');
      const result = row.libreoffice;
      if (result?.channel !== 'libreoffice' || result.version !== config.libreoffice_version)
        fail('LibreOffice comparison identity mismatch');
      if (result.status === 'failed') {
        if (!['capture', 'compare'].includes(result.stage) || !result.error?.trim() || 'penalized_ssim' in result)
          fail('invalid LibreOffice failure');
      } else {
        if (result.status !== 'ok') fail('invalid LibreOffice status');
        validateComparison(result);
        if (result.reference.sha256 !== row.source_sha256 ||
            result.reference_pages !== sample.metadata.reference.pages ||
            !Number.isInteger(result.actual_pages) || result.actual_pages < 1)
          fail('LibreOffice reference mismatch');
      }
      results.set(row.id, row);
    }
  }
  const samples = report.samples.map((sample) => {
    if (sample.format !== 'docx') return sample;
    const row = results.get(sample.id);
    if (!row) fail('missing sample');
    return { ...sample, comparisons: [...sample.comparisons, row.libreoffice], native_timings: row.timings };
  });
  timingSummary(samples);
  return { ...report, samples, docx_benchmark: JSON.parse(identity),
    docx_runners: inputs.map((part) => ({ shard: part.shard, ...part.runner })) };
}

export async function stageDocx(planPath, shard, output, cacheDir) {
  const bytes = await readFile(planPath);
  const plan = validatePlan(JSON.parse(bytes));
  const ids = docxShards(plan)[shard];
  if (!ids || !plan.docx_published_source_sha) fail('invalid shard or missing published source');
  const samples = ids.map((id) => plan.samples.find((sample) => sample.id === id));
  const tasks = [];
  for (const sample of samples) {
    const root = resolve(output, sample.id);
    await mkdir(resolve(root, 'reference'), { recursive: true });
    await writeFile(resolve(root, 'reference/result.json'), JSON.stringify(sample.metadata.reference));
    tasks.push({ sample, asset: sample.metadata.source, path: resolve(root, 'source.docx'), maximum: 128 * 1024 * 1024 });
    sample.metadata.reference_pages.forEach((asset, index) => tasks.push({
      sample, asset, path: resolve(root, 'reference', `page_${String(index + 1).padStart(4, '0')}.png`), maximum: 32 * 1024 * 1024,
    }));
  }
  let completed = 0;
  await mapPool(tasks, 8, async ({ sample, asset, path, maximum }) => {
    await writeFile(path, await fetchAsset(asset, sample.id, download, { cacheDir, origin: CORPUS_ORIGIN, maximum }));
    if (++completed % 50 === 0 || completed === tasks.length)
      console.log(`DOCX shard ${shard}: staged ${completed}/${tasks.length} verified assets`);
  });
  await writeFile(resolve(output, 'job.json'), JSON.stringify({
    plan_sha256: digest(bytes), shard, samples, source_sha: plan.source_sha,
    published_source_sha: plan.docx_published_source_sha, method: METHOD, trials: TRIALS,
    published_version: plan.versions.docx,
  }, null, 2) + '\n');
}

if (process.argv[1] && import.meta.url === pathToFileURL(resolve(process.argv[1])).href) {
  const { QUALITY_PLAN, QUALITY_OUTPUT, QUALITY_SHARD, QUALITY_ASSET_CACHE } = process.env;
  if (!QUALITY_PLAN || !QUALITY_OUTPUT || !/^\d+$/.test(QUALITY_SHARD ?? ''))
    throw new Error('QUALITY_PLAN, QUALITY_OUTPUT, and QUALITY_SHARD are required');
  await stageDocx(resolve(QUALITY_PLAN), Number(QUALITY_SHARD), resolve(QUALITY_OUTPUT), QUALITY_ASSET_CACHE);
}

File: scripts/office-quality/docx_benchmark.py
import hashlib
import json
import os
import platform
import re
import signal
import subprocess
import time
import xml.etree.ElementTree as ET
from pathlib import Path

import fitz
import numpy
import PIL
import skimage
from PIL import Image, ImageStat

from compare import Pages, compare

CHANNELS = ['published', 'commit', 'libreoffice']


def sha256(path):
    return hashlib.sha256(path.read_bytes()).hexdigest()


def write_json(path, value):
    path.write_text(json.dumps(value, indent=2, allow_nan=False) + '\n')


def execute(command, timeout=180, log=None):
    started = time.perf_counter_ns()
    with subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
                          text=True, start_new_session=True) as process:
        try:
            stdout, stderr = process.communicate(timeout=timeout)
        except subprocess.TimeoutExpired:
            os.killpg(process.pid, signal.SIGKILL)
            process.communicate()
            raise RuntimeError(f'Command timed out after {timeout}s: {command[0]}')
    elapsed = (time.perf_counter_ns() - started) / 1e6
    if log:
        log.write_text(stdout + stderr)
    if process.returncode:
        raise RuntimeError(f'{command[0]} exited {process.returncode}: {stderr[-4000:]}')
    return elapsed, stdout


def font_environment(manifest_path, output):
    manifest = json.loads(manifest_path.read_text())
    families = {}
    for face in manifest['faces']:
        file = manifest_path.parent / face['file']
        if file.parent != manifest_path.parent or sha256(file) != face['sha256']:
            raise ValueError('Bundled font hash mismatch')
        families[face['family'].lower()] = face['family']
        if face.get('metricCompatWith'):
            families[face['metricCompatWith'].lower()] = face['family']
    config = ET.Element('fontconfig')
    ET.SubElement(config, 'dir').text = str(manifest_path.parent)
    ET.SubElement(config, 'cachedir').text = str(output / 'font-cache')
    aliases = {**families, **{key: families[value] for key, value in manifest['aliases'].items() if value in families},
               'Calibri Light': 'Carlito', 'sans-serif': 'Liberation Sans',
               'serif': 'Liberation Serif', 'monospace': 'Liberation Mono'}
    for requested, substitute in sorted(aliases.items()):
        alias = ET.SubElement(config, 'alias', binding='strong')
        ET.SubElement(alias, 'family').text = requested
        ET.SubElement(ET.SubElement(alias, 'prefer'), 'family').text = substitute
    path = output / 'fonts.conf'
    ET.ElementTree(config).write(path, encoding='utf-8', xml_declaration=True)
    if platform.system() == 'Linux':
        os.environ['FONTCONFIG_FILE'] = str(path)
        execute(['fc-cache', '-f'])
    return sha256(manifest_path)


def lo_command(soffice, profile, output, source, filter_name):
    return [soffice, '-env:UserInstallation=' + profile.as_uri(), '--headless',
            '--nologo', '--nodefault', '--nofirststartwizard', '--convert-to',
            filter_name, '--outdir', str(output), str(source)]


def edge_adjust(image, target):
    if image.size == target:
        return image, None
    if any(abs(a - b) > 1 for a, b in zip(image.size, target)):
        return image, None
    adjustment = dict(original=list(image.size), target=list(target))
    edged = Image.new('RGB', target, 'white')
    edged.paste(image, (0, 0))
    return edged, adjustment


def libreoffice_fidelity(sample, root, soffice, version):
    format = sample.get('format', 'docx')
    actual = root / 'libreoffice'
    actual.mkdir()
    result = dict(channel='libreoffice', version=version)
    stage = 'capture'
    try:
        execute(lo_command(soffice, root / 'profile-fidelity', actual, root / f'source.{format}',
                           'pdf:impress_pdf_Export' if format == 'pptx' else 'pdf:writer_pdf_Export'), timeout=600, log=actual / 'convert.log')
        adjustments = []
        with fitz.open(actual / 'source.pdf') as pdf:
            if not len(pdf):
                raise ValueError('LibreOffice produced no pages')
            for index, page in enumerate(pdf):
                pixmap = page.get_pixmap(dpi=150, alpha=False)
                image = Image.frombytes('RGB', (pixmap.width, pixmap.height), pixmap.samples)
                reference = root / 'reference' / f'page_{index + 1:04d}.png'
                if reference.exists():
                    with Image.open(reference) as expected:
                        image, adjustment = edge_adjust(image, expected.size)
                    if adjustment:
                        adjustments.append(dict(page=index + 1, **adjustment))
                image.save(actual / f'page_{index + 1:04d}.png')
            write_json(actual / 'result.json', dict(status='ok', sha256=sample['metadata']['source']['sha256'],
                       dpi=150, pages=len(pdf), engine='LibreOffice ' + version,
                       rasterizer='PyMuPDF ' + fitz.VersionBind, edge_adjustments=adjustments))
        stage = 'compare'
        def progress(page, total):
            if page % 10 == 0 or page == total:
                print(f"{sample['id']}: scored {page}/{total} LibreOffice pages", flush=True)
        result.update(compare(Pages(root / 'reference', 150), Pages(actual, 150), root / 'lo-comparison',
                              gallery=False, progress=progress))
        result['status'] = 'ok'
    except Exception as error:
        result.update(status='failed', stage=stage, error=str(error))
    return result


def validate_png(path, size, require_ink):
    with Image.open(path) as image:
        image.load()
        if image.format != 'PNG' or any(abs(a - b) > 1 for a, b in zip(image.size, size)):
            raise ValueError(f'Invalid PNG dimensions: {image.size}, expected {size} ±1px')
        variance = ImageStat.Stat(image.convert('L')).var[0]
        if require_ink and variance < 0.1:
            raise ValueError('Renderer produced a blank first page')
        return dict(size=list(image.size), sha256=sha256(path), bytes=path.stat().st_size,
                    grayscale_variance=variance)


def measure_sample(sample, root, binaries, manifest, soffice, trials, offset=0):
    format = sample.get('format', 'docx')
    output = root / 'timing'
    output.mkdir()
    with Image.open(root / 'reference/page_0001.png') as reference:
        size = tuple(round(value * 96 / 150) for value in reference.size)
        require_ink = ImageStat.Stat(reference.convert('L')).var[0] >= 1
    options = {key: dict(type='long', value=value) for key, value in
               [('PixelWidth', size[0]), ('PixelHeight', size[1]), ('PageNumber', 1)]}
    options['Translucent'] = dict(type='boolean', value=False)
    commands = {channel: [str(binary), str(root / f'source.{format}'), str(output / f'{channel}.png'), str(manifest)]
                for channel, binary in binaries.items()}
    commands['libreoffice'] = lo_command(soffice, root / 'profile-timing', output, root / f'source.{format}',
                                       f'png:{"impress" if format == "pptx" else "writer"}_png_Export:' + json.dumps(options, separators=(',', ':')))
    paths = {channel: output / (f'{channel}.png' if channel != 'libreoffice' else 'source.png') for channel in CHANNELS}
    results = {channel: dict(status='ok', elapsed_ms=[], command=commands[channel]) for channel in CHANNELS}
    for repeat in range(trials + 1):
        rotation = (repeat + offset) % len(CHANNELS)
        order = CHANNELS[rotation:] + CHANNELS[:rotation]
        for channel in order:
            if results[channel]['status'] == 'failed':
                continue
            print(f"{sample['id']}: {channel} {'warmup' if repeat == 0 else f'trial {repeat}/{trials}'}", flush=True)
            try:
                paths[channel].unlink(missing_ok=True)
                elapsed, stdout = execute(commands[channel], log=output / f'{channel}-{repeat}.log')
                png = validate_png(paths[channel], size, require_ink)
                if channel != 'libreoffice':
                    metadata = json.loads(stdout)
                    if (metadata.get('status') != 'ok' or metadata.get('skipped_images') != 0
                            or metadata.get('source_sha256') != sample['metadata']['source']['sha256']
                            or metadata.get('png_sha256') != png['sha256'] or metadata.get('page') != 1
                            or metadata.get('dpi') != 96 or metadata.get('pages', 0) < 1):
                        raise ValueError('Native CLI returned invalid render metadata')
                    results[channel]['native'] = metadata
                results[channel]['png'] = png
                if repeat:
                    results[channel]['elapsed_ms'].append(elapsed)
                print(f"{sample['id']}: {channel} {elapsed:.1f} ms", flush=True)
            except Exception as error:
                results[channel] = dict(status='failed', error=str(error), command=commands[channel])
                print(f"{sample['id']}: {channel} FAILED: {error}", flush=True)
    return results


def main():
    root = Path(os.environ['QUALITY_OUTPUT']).resolve()
    job = json.loads((root / 'job.json').read_text())
    native = Path(os.environ['QUALITY_NATIVE']).resolve()
    manifest = Path(os.environ['QUALITY_FONTS']).resolve()
    soffice = os.environ.get('SOFFICE', 'libreoffice')
    os.environ['SAL_USE_VCLPLUGIN'] = 'svp'
    fonts_hash = font_environment(manifest, root)
    _, version_text = execute([soffice, '--version'])
    version = re.search(r'LibreOffice (\d+(?:\.\d+){2,3})\b', version_text)[1]
    builds, binaries = {}, {}
    for channel, source in [('published', job['published_source_sha']), ('commit', job['source_sha'])]:
        binary = native / channel / 'docx-native-bench'
        build = json.loads((binary.parent / 'build.json').read_text())
        if build['source_sha'] != source or sha256(binary) != build['binary_sha256']:
            raise ValueError('Native executable does not match the frozen source')
        builds[channel], binaries[channel] = build, binary
    report = dict(schema_version=1, plan_sha256=job['plan_sha256'], shard=job['shard'], samples=[],
                  benchmark=dict(method=job['method'], trials=job['trials'], dpi=96,
                                 published_version=job['published_version'],
                                 libreoffice_version=version, libreoffice_build=version_text.strip(),
                                 fonts_sha256=fonts_hash, builds=builds, environment=dict(
                                     os=platform.system(), architecture=platform.machine(),
                                     image_os=os.environ.get('ImageOS', 'local'),
                                     image_version=os.environ.get('ImageVersion', 'local'),
                                     python=platform.python_version(), pymupdf=fitz.VersionBind,
                                     numpy=numpy.__version__, pillow=PIL.__version__, skimage=skimage.__version__,
                                     font_policy='bundled-fontconfig' if platform.system() == 'Linux' else 'system-fonts')),
                  runner=dict(name=os.environ.get('RUNNER_NAME', platform.node()),
                              platform=platform.platform(), processors=os.cpu_count()))
    for index, sample in enumerate(job['samples']):
        directory = root / sample['id']
        if sha256(directory / 'source.docx') != sample['metadata']['source']['sha256']:
            raise ValueError('Source hash mismatch')
        for page, asset in enumerate(sample['metadata']['reference_pages'], 1):
            if sha256(directory / 'reference' / f'page_{page:04d}.png') != asset['sha256']:
                raise ValueError('Reference hash mismatch')
        print(f"DOCX shard {job['shard']}: {index + 1}/{len(job['samples'])} {sample['id']}", flush=True)
        timings = measure_sample(sample, directory, binaries, manifest, soffice, job['trials'], index)
        print(f"{sample['id']}: LibreOffice fidelity", flush=True)
        fidelity = libreoffice_fidelity(sample, directory, soffice, version)
        report['samples'].append(dict(id=sample['id'], source_sha256=sample['metadata']['source']['sha256'],
                                      libreoffice=fidelity, timings=timings))
        write_json(root / 'report.json', report)
        print(f"{sample['id']}: fidelity {fidelity['status']}, SSIM {fidelity.get('penalized_ssim', '—')}", flush=True)


if __name__ == '__main__':
    main()
